qa/rpc-tests/wallet_persistence.py

```diff
@@ -97,5 +98,34 @@ def run_test(self):
         assert_equal(self.nodes[0].z_getbalance(sapling_addr), Decimal('5'))
         assert_equal(self.nodes[1].z_getbalance(dest_addr), Decimal('15'))
 
+        # Verify importing a spending key will update and persist the nullifiers and witnesses correctly
+        sk0 = self.nodes[0].z_exportkey(sapling_addr)
+        self.nodes[2].z_importkey(sk0, "yes")
+        assert_equal(self.nodes[2].z_getbalance(sapling_addr), Decimal('5'))
+
+        # Restart the nodes
+        stop_nodes(self.nodes)
+        wait_bitcoinds()
+        self.setup_network()
+
+        # Verify nullifiers persisted correctly by checking balance
+        # Prior to PR #3590, there will be an error as spent notes are considered unspent:
+        #    Assertion failed: expected: <25.00000000> but was: <5>
```

Comment:
Is the message in the comment correct? It seems the wrong way round.

Reply:
The comment is correct  (copied direct). It's a nit due to the order of the parameters to assert_equal.  The first parameter is what is expected, and the second is what it is.  https://github.com/zcash/zcash/blob/379c7439a0b8028caf9fbb4a47caed9da2434193/qa/rpc-tests/test_framework/util.py#L358
It seems there is bias on the part of the programmers so far to pass parameters the other way; it may be a visual thing, to have: `assert_equal(lotsofstuffblahblahblah, 3.14)` where it is easier to identify and edit the expected value at the tned of the line.

Reply:
I think I changed this at one point to allow passing in a message. I mimicked what JUnit does which is why it has the expected/actual. Having done this, I think I now prefer the way rust does it - just showing left and right rather than "expected" and "actual". It would be nice to change this since probably half of the tests do not follow this convention.